src/sources/models.js

```diff
@@ -3,14 +3,67 @@ const {fetch} = require("../fetch");
 const {NoResourcePathError} = require("../exceptions");
 const utils = require("../utils");
 
-
-/* These Source, Dataset, and Narrative classes contain information to map an
- * array of dataset/narrative path parts onto a URL.  Source selection and
- * dataset path aliasing (/flu → /flu/seasonal/h3n2/ha/3y) is handled in
- * utils/prefix.parsePrefix().
+/* The model classes here are the base classes for the classes defined in
+ * ./core.js, ./community.js, ./groups.js, etc.
+ *
+ * Sources represent a remote HTTP data source.  Sources contain Datasets and
+ * Narratives (both Resources).  Datasets and Narratives have Subresources
+ * representing specific files/data that make up the conceptual Resource as a
+ * whole.
+ *
+ * Source
+ *   Dataset                (implements Resource interface)
+ *     DatasetSubresource   (implements Subresource interface)
+ *   Narrative              (implements Resource interface)
+ *     NarrativeSubresource (implements Subresource interface)
+ *
+ * Subresources are separate from Resources so that our code can more easily
+ * talk about, address, and pass around values representing both the conceptual
+ * whole of a dataset or narrative and the concrete individual files making up
+ * those wholes.
+ *
+ * A concrete example:
+ *
+ * CoreSource (in ./core.js) represents a Cloudfront distribution
+ * (https://data.nextstrain.org) in front of an S3 bucket
+ * (s3://nextstrain-data).
+ *
+ *   const coreSource = new CoreSource()
+ *
+ * The core dataset "flu/seasonal/h3n2/ha/2y" is represented by a Dataset
+ * instance you get by calling:
+ *
+ *   const dataset = coreSource.dataset(["flu", "seasonal", "h3n2", "ha", "2y"])
+ *
+ * That Dataset has Subresources identified by the names "main" and
+ * "tip-frequencies", which you get by calling:
+ *
+ *   dataset.subresource("main")
+ *   dataset.subresource("tip-frequencies")
+ *
+ * These Subresources can be retrieved at the following URLs, which you obtain
```

Comment:
Do you foresee these always being HTTP endpoints? I am thinking in particular of things backed by S3 buckets, where we currently use `S3.getSignedUrl`. Unsure if there are speed penalties / limited functionality when we use signed URLs (vs, e.g. a `S3.getObject` call) 

Guaranteeing that it's always HTTP is nice in that we can run `fetch(source.dataset.subresource(type).url())` without knowing anything about how things are stored behind the scenes, but I could imagine another abstraction where we ran `source.dataset.getSubresouce(type)` and it handled how data is fetched.

Reply:
I think they can likely remain HTTP endpoints ~indefinitely, and there are advantages to doing so. That said, we could add support for other URL schemes/wire protocols in the future if we needed and still handle them pretty transparently within this codebase.

Good to think about limitations of things, but I don't think we have much to worry about with S3 signed URLs. They're basically exposing what S3 clients are doing under the hood normally to make requests. While S3 client APIs present the facade of a "session", behind the facade it's all separately authenticated HTTP requests that look just like signed URLs. So the distinction of using signed URLs vs. not is mostly whether you want to talk plain HTTP or use a client library's API instead. There's a few things which if you need them are a little more ceremony without the benefit of a client library, like multi-part uploads, but nothing that's not doable.

The advantages of URLs over an already-loaded data structure or even an IO stream returned by something like `source.dataset.getSubresouce(type)` are:

- You can implement the latter (loaded data structure or IO streams) with the former (URLs), but not vice versa. It's a more basal concept.
- URLs and HTTP are a looser coupling between components (e.g. easier to change one component later without affecting the other).
- We can re-use the design decisions already made by the HTTP spec for how to transmit metadata about some data stream (i.e. so we don't have to re-invent this)
- The opportunities for optimization are greater, since we're mostly making these HTTP requests to service our own HTTP responses. For example, one optimization might be to dynamically choose to redirect to the subresource URL if the upstream Source supports CORs instead of always proxying the data through us. (I've had this roughly prototyped locally for a bit.)

Reply:
Added a slightly revised version of that last part to the comment.